### PR TITLE
Replace SAO implementation with MSSAO

### LINKED ISSUE #46: Improve / rework SSAO
**Describe the project you are working on:**
Some simple 3d project.

**Describe how this feature / enhancement will help your project:**
Better visual quality, aswell as SSAO looking more like what you'd expect from it. It may also be a performance increase

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
This is a test implementation (not in godot) at the lowest usable quality in my case, which is 3 samples per pixel at 1/4* resolution. (Edit: I was able to push the quality a bit more by optimizing the sample offsets)
\* Edit2: note that by *1/4* I mean half the width and height, so a quarter of the area / #pixels.
grafik
grafik

In comparison, at the lowest setting godot uses **8 samples per pixel at full resolution** which looks like this:
grafik

Even on higher settings this doesn't improve much and isn't worth the performance impact in my opinion.

**Describe implementation detail for your proposal (in code), if possible:**
Using better sample distributions and filtering it's possible to improve the quality of SSAO a lot.
**EDIT:** I will upload a reference implementation [here](https://github.com/CptPotato/SSAO)

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
no

**Is there a reason why this should be core and not an add-on in the asset library?:**
As it is an improvement of existing core functionality it doesn't make much sense to provide it as an addon.

**Further notes:**
I'd target this for 4.x or 4.0. I'm not sure if I'll have the time to integrate this into godot but providing a reference implementation shouldn't be a problem.

### PR DESCRIPTION
**Pull request by clayjohn**
Closes https://github.com/godotengine/godot-proposals/issues/46

**What is this?**
Mostly a direct port from https://github.com/microsoft/DirectX-Graphics-Samples/tree/mast/MiniEngine but I have had to make many small changes to port to Vulkan/work with Godot style and priorities.

Miniengine AO itself is based on "Multi scale ambient occlusion" (MSSAO) and "Volumetic Obscurance" (VAO).

**Why use Miniengine AO?**
The reason this approach works so well is that it heavily optimized for compute shaders. This means it is incredibly fast and high-quality on compute-enabled devices. Unfortunately, it cannot be ported to lower end rendering backends. We will have to use a simpler method when the time comes. 

This is the best approach for Godot because it is the only approach that doesn't rely heavily on TAA. Right now the most popular SSAO technique is "ground truth ambient occlusion" (GTAO). GTAO is used in Unreal for example. GTAO looks amazing and is incredibly fast, however, without TAA it looks very grainy and you would see the noise from the low sample count. What makes Miniengine AO so good is that it doesn't rely on stochastic sampling and so it always looks good.

**How it works**
From a high level, the approach is to downsample the depth buffer 4 times creating four levels. Then compute AO at each level. Then starting from the smallest mipmap, blur and upsample the AO using the next smallest mipmap. This way AO is computed over multiple resolutions (which ends up meaning distances) and is continuously blended as it is upsampled.

**Downsides**
There are two issues with it right now. The first is a side effect of getting large scale AO and it is that this AO is more prone to flickering when you have objects close together and the camera is moving fast (like the columns in sponza below). ~~The second is barely noticeable in the pictures below, but there is a slight darkening at the edges of the screen. I have some ideas for how to reduce it, but they can wait for a subsequent PR.~~ _edit: fixed :)_

**Future**
If there is need, I think it is possible to compute at half or quarter resolution. However, it is already very fast. One my hardware, low quality takes about 0.25ms while default quality takes about 0.35 (at full 1920x1080).

*Before: 2-4ms per frame, highest quality setting*
sao-hq-sponza

*After: 0.7ms per frame, ultra quality setting*
mssao-hq-sponza

*Before*
sponza-sao

*After*
sponza-mssao

### CODE REVIEW

File: doc/classes/RenderingServer.xml [hunk 1]
```diff
@@ -773,21 +773,24 @@
...
+			<argument index="7" name="levels" type="int">
 			</argument>
-			<argument index="7" name="blur" type="int" enum="RenderingServer.EnvironmentSSAOBlur">
+			<argument index="8" name="light_affect" type="float">
 			</argument>
-			<argument index="8" name="bilateral_sharpness" type="float">
+			<argument index="9" name="ao_channel_affect" type="float">
 			</argument>
 			<description>
+				Sets the variables to be used with the SSAO effect. Godot uses a form of MSSAO based on Microsofts MiniEngine demo. This form of SSAO samples SSAO at multiple levels of decreasing resolution in order to compute SSAO over large distances. See [Environment] for more details and specific usage.
```
Comment (akien-mga) [inline doc/classes/RenderingServer.xml:793]:
```suggestion
				Sets the variables to be used with the SSAO effect. Godot uses a form of MSSAO based on Microsoft's MiniEngine demo. This form of SSAO samples SSAO at multiple levels of decreasing resolution in order to compute SSAO over large distances. See [Environment] for more details and specific usage.
```

File: scene/resources/environment.cpp [hunk 2]
```diff
@@ -346,13 +346,33 @@ bool Environment::is_ssao_enabled() const {
...
+	ssao_noise_tolerance = p_noise_tolerance;
 	_update_ssao();
 }
-
-float Environment::get_ssao_radius() const {
-	return ssao_radius;
+float Environment::get_ssao_noise_tolerance() const {
+	return ssao_noise_tolerance;
+}
+void Environment::set_ssao_blur_tolerance(float p_blur_tolerance) {
```
Comment (akien-mga) [inline scene/resources/environment.cpp:356]: Add a line between all method definitions.

File: servers/rendering/rasterizer_rd/rasterizer_effects_rd.cpp [hunk 3]
```diff
@@ -926,156 +926,337 @@ void RasterizerEffectsRD::bokeh_dof(RID p_base_texture, RID p_depth_texture, con
...
+	if (!p_use_full_samples) {
+		ssao.ssao_render_push_constant.sample_weight_table[0] = 0.0f;
+		ssao.ssao_render_push_constant.sample_weight_table[2] = 0.0f;
+		ssao.ssao_render_push_constant.sample_weight_table[5] = 0.0f;
+		ssao.ssao_render_push_constant.sample_weight_table[7] = 0.0f;
+		ssao.ssao_render_push_constant.sample_weight_table[9] = 0.0f;
+	}
+
+	// Normalize the weights by dividing by the sum of all weights
+	float totalWeight = 0.0f;
```
Comment (akien-mga) [inline servers/rendering/rasterizer_rd/rasterizer_effects_rd.cpp:1014]:
```suggestion
	float total_weight = 0.0f;
```
(Unless it's a direct copy from the MS code and you prefer to keep it similar.)

File: servers/rendering/rasterizer_rd/rasterizer_effects_rd.cpp [hunk 4]
```diff
@@ -926,156 +926,337 @@ void RasterizerEffectsRD::bokeh_dof(RID p_base_texture, RID p_depth_texture, con
...
-	ssao.minify_push_constant.orthogonal = p_projection.is_orthogonal();
-	ssao.minify_push_constant.z_near = p_projection.get_z_near();
-	ssao.minify_push_constant.z_far = p_projection.get_z_far();
-	ssao.minify_push_constant.pixel_size[0] = 1.0 / p_depth_buffer_size.x;
-	ssao.minify_push_constant.pixel_size[1] = 1.0 / p_depth_buffer_size.y;
-	ssao.minify_push_constant.source_size[0] = p_depth_buffer_size.x;
-	ssao.minify_push_constant.source_size[1] = p_depth_buffer_size.y;
-
-	RD::ComputeListID compute_list = RD::get_singleton()->compute_list_begin();
+void RasterizerEffectsRD::_compute_ssao(RD::ComputeListID compute_list, RID p_destination, RID p_depth_buffer, const float p_tan_half_fov_h, int p_width, int p_height, int p_depth, float p_rejection_radius, float p_intensity, bool p_use_full_samples) {
```
Comment (akien-mga) [inline servers/rendering/rasterizer_rd/rasterizer_effects_rd.cpp:929]:
```suggestion
void RasterizerEffectsRD::_compute_ssao(RD::ComputeListID p_compute_list, RID p_destination, RID p_depth_buffer, const float p_tan_half_fov_h, int p_width, int p_height, int p_depth, float p_rejection_radius, float p_intensity, bool p_use_full_samples) {
```

File: servers/rendering/rasterizer_rd/rasterizer_effects_rd.cpp [hunk 5]
```diff
@@ -926,156 +926,337 @@ void RasterizerEffectsRD::bokeh_dof(RID p_base_texture, RID p_depth_texture, con
...
+	const float sample_thickness[12] = {
```
Comment (akien-mga) [inline servers/rendering/rasterizer_rd/rasterizer_effects_rd.cpp:930]: Maybe add a link to the original algorithm implementation as reference? Here or wherever it's most appropriate.

File: servers/rendering/rasterizer_rd/shaders/ssao_downsample1.glsl [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	imageStore(DS2x, ivec2(st), vec4(w1));
+	imageStore(DS2xAtlas, ivec3(st >> 2, slice), vec4(w1));
+
+	if ((GI & 011) == 0) {
+		st = DTid.xy >> 1;
+		slice = (st.x & 3) | ((st.y & 3) << 2);
+		imageStore(DS4x, ivec2(st), vec4(w1));
+		imageStore(DS4xAtlas, ivec3(st >> 2, slice), vec4(w1));
+	}
+}
```
Comment (akien-mga) [inline servers/rendering/rasterizer_rd/shaders/ssao_downsample1.glsl:69]: Missing `\n`.

File: servers/rendering/rasterizer_rd/shaders/ssao_downsample2.glsl [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	imageStore(DS8xAtlas, ivec3(stAtlas, stSlice), w1);
+
+	if ((GI & 011) == 0) {
+		uvec2 st = DTid.xy >> 1;
+		uvec2 stAtlas = st >> 2;
+		uint stSlice = (st.x & 3) | ((st.y & 3) << 2);
+		imageStore(DS16x, ivec2(st), w1);
+		imageStore(DS16xAtlas, ivec3(stAtlas, stSlice), w1);
+	}
+}
```
Comment (akien-mga) [inline servers/rendering/rasterizer_rd/shaders/ssao_downsample2.glsl:42]: Missing `\n`.

File: servers/rendering/rasterizer_rd/shaders/ssao_downsample1.glsl (hunk 6 above)
Comment (akien-mga) [reply servers/rendering/rasterizer_rd/shaders/ssao_downsample1.glsl:69]: Same for other new `.glsl` files.

**Comment by clayjohn** [comment]
@zireael07 unfortunately no, this form of SSAO requires compute shaders.

File: scene/resources/environment.cpp [hunk 8]
```diff
@@ -346,13 +346,40 @@ bool Environment::is_ssao_enabled() const {
...
+float Environment::get_ssao_upsample_tolerance() const {
+	return ssao_upsample_tolerance;
+}
+
+void Environment::set_ssao_rejection_radius(float p_rejection_radius) {
+	ssao_rejection_radius = p_rejection_radius;
+	_update_ssao();
+}
+
+float Environment::get_ssao_rejection_radius() const {
```
Comment (reduz) [inline scene/resources/environment.cpp:381]: Is it very common to modify these values? If the defaults work fine for say 99% of cases, I suggest moving them to project settings.

**Review by reduz — APPROVED** [review]

**Comment by reduz** [comment]
There are some issues with this, not entirely sure if its a problem of the algorithm or the implementation, for which edges retain size in screen space rather than scaling with depth.

mssao_size
mssao_size3

Hopefully its an implementation problem and not an algorithm problem so it can be figured out.

**Comment by GoldenThumbs** [comment]
@reduz Couldn't one do something like `radius/position.z`?

**Comment by clayjohn** [comment]
For future reference, we have decided not to go with MSSAO. While it is super fast and looks great in still images, it has severe issues in motion (as shown in reduz' gifs above). 

We will look into going with an HBAO-based solution like [ASSAO](https://github.com/GameTechDev/ASSAO) from intel or [CACAO](https://github.com/GPUOpen-Effects/FidelityFX-CACAO) from AMD.

**Comment by GoldenThumbs** [comment]
> For future reference, we have decided not to go with MSSAO. While it is super fast and looks great in still images, it has severe issues in motion (as shown in reduz' gifs above).
> 
> We will look into going with an HBAO-based solution like [ASSAO](https://github.com/GameTechDev/ASSAO) from intel or [CACAO](https://github.com/GPUOpen-Effects/FidelityFX-CACAO) from AMD.

This issue should be easily fixed by changing line `237` in `ssao.glsl` to `float ssDiskRadius = -params.proj_scale * params.radius / C.z`, has this been tried yet? I'm know I posted about this before but those gifs *really* look like they stem from this.

**Comment by clayjohn** [comment]
@GoldenThumbs the code you point to is in the old SSAO code. 

That being said, if you can figure out how to "fix" the MSSAO implementation, please do. 

I tried a few variations of trying to scale the radius in world space, but I had no luck. I have a feeling it isn't possible without substantially altering the approach.

**Comment by GoldenThumbs** [comment]
@clayjohn I checked, it's missing the `/ C.z`. There could be other issues here that lead to a similar artifact.

**Comment by clayjohn** [comment]
@GoldenThumbs yes. The C.z is missing. But you are looking at code that was deleted in this PR. It is not MSSAO code, it is SAO, which doesn't have that same artifact.

**Comment by GoldenThumbs** [comment]
Ok, I see that now. Sorry about that! I'll do more reading into MSSAO.

**Pull request merged**